### PR TITLE
[Feature 18] implemented backend for the account management feature

### LINKED ISSUE #18: [Feature 18] Account Management
backend implementation to change account details.

### LINKED ISSUE #27: [Task 3] Create Tests and Documentation for EditAccount
I'm going to do this later, once the implementation of account is pulled.

### PR DESCRIPTION
**Pull request by ABunchOfAtom**
Created account editing controllers and use cases.

### CODE REVIEW

**Comment by jenniferjinyoungyang** [comment]
Can you add the unit testing for the controller and interactor? And android testing for the database-functions?

**Comment by jenniferjinyoungyang** [comment]
email should be unique. Please make the updateEmail method can handle this case i.e. check if there is another account using the same email, if so, then it should be handled appropriately.

**Review by jenniferjinyoungyang — CHANGES_REQUESTED** [review]
I left the comments above.

File: app/src/main/java/com/courseproject/tindar/entities/Account.java
Comment (navarhontes) [inline app/src/main/java/com/courseproject/tindar/entities/Account.java:1]: Are we using this class at all? Are we planning to use this class at all? If we keep this file then I'd comment the firstname/lastname stuff out, rather than handling it this way.

**Review by navarhontes — DISMISSED** [review]
Seconding Jennifer's comments on JavaDocs and unit testing. If you don't want to write full JavaDocs right now, then at least write a more descriptive PR to make review easier for us now and JavaDocs easier for yourself later, and make an issue. 

I've spent 20 minutes trying to make sense of the design choices here; not sure if it's just me or if there's an actual issues. Also very confused because EditAccountInputBoundary and EditAccountDsGateway look identical to me; why do they both exist at the same time? I think it would be helpful for both your and our sakes for you to write JavaDocs or at least a detailed PR comment about your design choices here.

File: app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java [hunk 1]
```diff
@@ -149,6 +152,59 @@ public String addAccount(boolean isActiveStatus, String email, String password,
...
+        SQLiteDatabase db = this.getWritableDatabase();
+        ContentValues cv = new ContentValues();
+        cv.put(IS_ACTIVE_STATUS, isActiveStatus);
+
+        db.update(TABLE_ACCOUNTS, cv, ID + "=?", new String[]{userId});
+        db.close();
+    }
+
+    @Override
+    public void updateEmail(String userId, String email) {
```
[the remark this answers is not in the document]
Comment (A-Bunch-Of-Atoms) [reply app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java]: done

File: app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java [hunk 2]
```diff
@@ -149,6 +152,59 @@ public String addAccount(boolean isActiveStatus, String email, String password,
...
+        SQLiteDatabase db = this.getWritableDatabase();
+        ContentValues cv = new ContentValues();
+        cv.put(EMAIL, email);
+
+        db.update(TABLE_ACCOUNTS, cv, ID + "=?", new String[]{userId});
+        db.close();
+    }
+
+    @Override
```
[the remark this answers is not in the document]
Comment (A-Bunch-Of-Atoms) [reply app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java:242]: done

**Comment by A-Bunch-Of-Atoms** [comment]
Do not review this yet. Still need to finish database tests and interactor tests.

**Comment by jenniferjinyoungyang** [comment]
You have commits from the main, and changes from those commits are displayed as your changes. It is hard to review like this. Can you clean out the commits history?

**Review by kresimirgotovac — COMMENTED** [review]
the documentation you have done looks clean, just need to add to the places you've missed i.e database helper methods

**Review by kresimirgotovac — DISMISSED** [review]
Approving because all substantial comments have been made already

**Comment by jenniferjinyoungyang** [comment]
Can you please write unit tests for the EditAccountInteractor?

Also, for the unit tests for EditAccountController, can you pls cover the case for the empty password for the updatePassword method?
image

**Comment by jenniferjinyoungyang** [comment]
pls resolve merge conflict also.

**Review by jenniferjinyoungyang — CHANGES_REQUESTED** [review]
left comments above

**Comment by A-Bunch-Of-Atoms** [comment]
Fixed merge conflicts.

**Review by kresimirgotovac — APPROVED** [review]

**Review by kresimirgotovac — DISMISSED** [review]

File: app/src/androidTest/java/com/courseproject/tindar/ds/DatabaseHelperTest.java [hunk 3]
```diff
@@ -60,6 +59,8 @@ public void setUp() {
 
     @After
     public void tearDown() {
+        dbHelper.deleteLikeLists();
+        dbHelper.deleteAccounts();
```
Comment (jenniferjinyoungyang) [inline app/src/androidTest/java/com/courseproject/tindar/ds/DatabaseHelperTest.java:63]: these two lines no longer needed and should be deleted. Now we have dbHelper.deleteAllDbRecords() inside setup to clean out the db records

File: app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java [hunk 4]
```diff
@@ -239,6 +242,99 @@ public boolean checkIfEmailAlreadyUsed(String email) {
...
+        SQLiteDatabase db = this.getWritableDatabase();
+        ContentValues cv = new ContentValues();
+        cv.put(PASSWORD, password);
+
+        db.update(TABLE_ACCOUNTS, cv, ID + "=?", new String[]{userId});
+        db.close();
+        return true;
+    }
+
+    public void deleteAccounts() {
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java:328]: this method is no longer needed and should be deleted.

File: app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java [hunk 5]
```diff
@@ -239,6 +242,99 @@ public boolean checkIfEmailAlreadyUsed(String email) {
...
+        SQLiteDatabase db = this.getWritableDatabase();
+        db.execSQL("delete from "+ TABLE_ACCOUNTS);
+    }
+
+    public void deleteLikeLists() {
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java:333]: this method is no longer needed and should be deleted.

**Review by kresimirgotovac — DISMISSED** [review]

File: app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java (hunk 5 above)
Comment (A-Bunch-Of-Atoms) [reply app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java]: done

File: app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java (hunk 4 above)
Comment (A-Bunch-Of-Atoms) [reply app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java]: done

File: app/src/androidTest/java/com/courseproject/tindar/ds/DatabaseHelperTest.java (hunk 3 above)
Comment (A-Bunch-Of-Atoms) [reply app/src/androidTest/java/com/courseproject/tindar/ds/DatabaseHelperTest.java]: done

**Review by jenniferjinyoungyang — DISMISSED** [review]

**Review by kresimirgotovac — DISMISSED** [review]

**Review by yiwxng — DISMISSED** [review]

**Review by jenniferjinyoungyang — DISMISSED** [review]

**Review by kresimirgotovac — DISMISSED** [review]

**Review by yiwxng — DISMISSED** [review]

**Review by jenniferjinyoungyang — APPROVED** [review]

**Review by candidcalcifer — APPROVED** [review]

**Review by yiwxng — APPROVED** [review]
